File: fsharp/Sieve.fs
namespace Benchmarks

open System

type Sieve() =
    inherit Benchmark()

    let mutable limit = 0L
    let mutable checksum = 0u

    override this.Prepare() =
        limit <- this.ConfigVal("limit")
        checksum <- 0u

    override this.Name = "Etc::Sieve"

    override this.Checksum = checksum

    override this.Run(_: int64) =
        let lim = int limit
        let primes = Array.create (lim + 1) 1uy
        primes.[0] <- 0uy
        primes.[1] <- 0uy

        let sqrtLimit = int (sqrt (float lim))

        for p in 2 .. sqrtLimit do
            if primes.[p] = 1uy then
                let mutable multiple = p * p
                while multiple <= lim do
                    primes.[multiple] <- 0uy
                    multiple <- multiple + p

        let mutable lastPrime = 2
        let mutable count = 1

        let mutable n = 3
        while n <= lim do
            if primes.[n] = 1uy then
                lastPrime <- n
                count <- count + 1
            n <- n + 2

        checksum <- checksum + (uint32 (lastPrime + count))

File: swift/Sources/benchmarks/Sieve.swift
import Foundation

final class Sieve: BenchmarkProtocol {
  private var limitVal: Int64 = 0
  private var checksumVal: UInt32 = 0

  init() {
    limitVal = configValue("limit") ?? 0
  }

  func run(iterationId: Int) {
    let limit = Int(limitVal)
    var primes = [UInt8](repeating: 1, count: limit + 1)
    primes[0] = 0
    primes[1] = 0

    let sqrtLimit = Int(sqrt(Double(limit)))

    for p in 2...sqrtLimit {
      if primes[p] == 1 {
        var multiple = p * p
        while multiple <= limit {
          primes[multiple] = 0
          multiple += p
        }
      }
    }

    var lastPrime = 2
    var count = 1

    var n = 3
    while n <= limit {
      if primes[n] == 1 {
        lastPrime = n
        count += 1
      }
      n += 2
    }

    checksumVal &+= UInt32(lastPrime + count)
  }

  var checksum: UInt32 {
    return checksumVal
  }

  func prepare() {}

  func name() -> String {
    return "Etc::Sieve"
  }
}
